From: Tejun Heo <teheo@suse.de>
Subject: [PATCH] ahci: disable PMP support on SB600/700
References: 299010

SB600 claims to support PMP but the controller locks up if SRST with
PMP==15 is issued.  Disable PMP support on SB600.  Whether SB700 is
affected by this problem is unclear, disable PMP support on SB700 too
for now.

Signed-off-by: Tejun Heo <teheo@suse.de>
---
 drivers/ata/ahci.c |   11 +++++++++--
 1 file changed, 9 insertions(+), 2 deletions(-)

--- a/drivers/ata/ahci.c	2007-08-27 14:01:23.000000000 -0400
+++ b/drivers/ata/ahci.c	2007-08-27 14:01:23.000000000 -0400
@@ -178,6 +178,7 @@ enum {
 	AHCI_HFLAG_32BIT_ONLY		= (1 << 4), /* force 32bit */
 	AHCI_HFLAG_MV_PATA		= (1 << 5), /* PATA port */
 	AHCI_HFLAG_NO_MSI		= (1 << 6), /* no PCI MSI */
+	AHCI_HFLAG_NO_PMP		= (1 << 7), /* no PMP */
 
 	/* ap->flags bits */
 
@@ -373,7 +374,7 @@ static const struct ata_port_info ahci_p
 	},
 	/* board_ahci_vt8251 */
 	{
-		AHCI_HFLAGS	(AHCI_HFLAG_NO_NCQ),
+		AHCI_HFLAGS	(AHCI_HFLAG_NO_NCQ | AHCI_HFLAG_NO_PMP),
 		.flags		= AHCI_FLAG_COMMON,
 		.link_flags	= AHCI_LFLAG_COMMON | ATA_LFLAG_HRST_TO_RESUME,
 		.pio_mask	= 0x1f, /* pio0-4 */
@@ -392,7 +393,7 @@ static const struct ata_port_info ahci_p
 	/* board_ahci_sb600 */
 	{
 		AHCI_HFLAGS	(AHCI_HFLAG_IGN_SERR_INTERNAL |
-				 AHCI_HFLAG_32BIT_ONLY),
+				 AHCI_HFLAG_32BIT_ONLY | AHCI_HFLAG_NO_PMP),
 		.flags		= AHCI_FLAG_COMMON,
 		.link_flags	= AHCI_LFLAG_COMMON,
 		.pio_mask	= 0x1f, /* pio0-4 */
@@ -590,6 +591,12 @@ static void ahci_save_initial_config(str
 		cap &= ~HOST_CAP_NCQ;
 	}
 
+	if ((cap && HOST_CAP_PMP) && (hpriv->flags & AHCI_HFLAG_NO_PMP)) {
+		dev_printk(KERN_INFO, &pdev->dev,
+			   "controller can't do PMP, turning off CAP_PMP\n");
+		cap &= ~HOST_CAP_PMP;
+	}
+
 	/* fixup zero port_map */
 	if (!port_map) {
 		port_map = (1 << ahci_nr_ports(cap)) - 1;
